lib/fuser_basic.h

```diff
@@ -150,17 +188,66 @@ struct BasicGateFuser final {
 
         gates_fused.push_back(std::move(gate_f));
       }
+
+      if (delayed_measurement_gate != nullptr) {
+        auto pgate = delayed_measurement_gate;
+
+        const auto& mea_gates_at_time = measurement_gates[pgate->time];
+
+        GateFused gate_f = {pgate->kind, pgate->time, 0, {}, pgate, {}};
+
+        // Fuse measurement gates with equal times.
+
+        for (const auto* pgate : mea_gates_at_time) {
+          gate_f.num_qubits += pgate->num_qubits;
+          gate_f.qubits.insert(gate_f.qubits.end(),
+                               pgate->qubits.begin(), pgate->qubits.end());
+        }
+
+        gates_fused.push_back(std::move(gate_f));
+      }
     }
 
     return gates_fused;
   }
 
  private:
+  static std::vector<unsigned> MergeWithMeasurementTimes(
+      const std::vector<Gate>& gates, const std::vector<unsigned>& times) {
+    std::vector<unsigned> times2;
+    times2.reserve(gates.size() + times.size());
+
+    std::size_t last = 0;
+
+    for (const auto& gate : gates) {
+      if (gate.kind == gate::kMeasurement && times2.back() < gate.time) {
+        times2.push_back(gate.time);
+      }
+
+      if (last < times.size() && gate.time > times[last]) {
+        while (last < times.size() && times[last] <= gate.time) {
+          unsigned prev = times[last++];
+          times2.push_back(prev);
+          while (last < times.size() && times[last] <= prev) ++last;
+        }
+      }
+
+      if (last == times.size()) break;
+    }
+
+    if (last < times.size()) {
+      times2.push_back(times[last]);
+    }
```

Comment:
Is it possible that this could omit some of `times` from the returned value? Do we want this to happen?

Reply:
Yes, it is possible if all the remaining `times` are greater than `gates.back().time`. We don't need them.